image_view/src/nodes/image_saver.cpp

```diff
@@ -38,35 +38,128 @@
 #include <image_transport/image_transport.h>
 #include <camera_calibration_parsers/parse.h>
 #include <boost/format.hpp>
+#include <thread>
+#include <mutex>
 
 #include <std_srvs/Empty.h>
 #include <std_srvs/Trigger.h>
 
-boost::format g_format;
-bool save_all_image, save_image_service;
-std::string encoding;
-bool request_start_end;
-
-
-bool service(std_srvs::Empty::Request &req, std_srvs::Empty::Response &res) {
-  save_image_service = true;
-  return true;
-}
 
 /** Class to deal with which callback to call whether we have CameraInfo or not
  */
 class Callbacks {
 public:
-  Callbacks() : is_first_image_(true), has_camera_info_(false), count_(0) {
+  Callbacks(ros::NodeHandle& nh,
+            const std::string& topic,
+            const std::string& filename_format,
+            const std::string& encoding,
+            const bool save_all_image,
+            const bool wait_for_save,
+            const double save_timeout)
+      : nh_(nh), it_(nh_), topic_(topic), encoding_(encoding),
+        count_(0), start_time_(0),
+        subscribing_(false), should_unsubscribe_(false),
+        wait_for_save_(wait_for_save), save_timeout_(save_timeout)
+  {
+    format_.parse(filename_format);
+    if (save_all_image)
+    {
+      subscribe();
+    } else {
+      srv_save_ = nh_.advertiseService("save", &Callbacks::callbackSave, this);
+      srv_start_ = nh_.advertiseService("start", &Callbacks::callbackStartSave, this);
+      srv_end_ = nh_.advertiseService("end", &Callbacks::callbackEndSave, this);
+    }
   }
 
-  bool callbackStartSave(std_srvs::Trigger::Request &req,
-                         std_srvs::Trigger::Response &res)
+  void subscribe()
   {
-    ROS_INFO("Received start saving request");
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    if (subscribing_) return;
+    subscribing_ = true;
+
+    ROS_DEBUG("subscribe");
+
+    is_first_image_ = true;
+    has_camera_info_ = false;
+
     start_time_ = ros::Time::now();
     end_time_ = ros::Time(0);
 
+    // Useful when CameraInfo is being published
+    sub_camera_ = it_.subscribeCamera(topic_, 1, &Callbacks::callbackWithCameraInfo, this);
+    // Useful when CameraInfo is not being published
+    sub_image_ = it_.subscribe(topic_, 1, &Callbacks::callbackWithoutCameraInfo, this);
+  }
+
+  void unsubscribe()
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    if (!subscribing_) return;
+    subscribing_ = false;
+
+    ROS_DEBUG("unsubcribe");
+
+    start_time_ = ros::Time(0);
+    end_time_ = ros::Time::now();
+    should_unsubscribe_ = false;
+
+    sub_camera_.shutdown();
+    sub_image_.shutdown();
+  }
+
+  void callbackTimeoutTimer(const ros::TimerEvent&)
+  {
+    ROS_DEBUG("callback timeout timer");
+    unsubscribe();
+  }
+
+  bool callbackSave(std_srvs::Empty::Request &req,
+                    std_srvs::Empty::Response &res)
+  {
+    if (subscribing_)
```

Comment:
If the `std::lock_guard` above is to guard this variable, you should do it here as well. Locking the mutex only while writing doesn't really guard it because reading the variable here then becomes a non-atomic operation.